pkg/distributor/query.go

```diff
@@ -249,15 +267,15 @@ func (d *Distributor) queryIngesterStream(ctx context.Context, replicationSets [
 			return ingesterQueryResult{}, err
 		}
 
-		result := ingesterQueryResult{}
-
 		// Why retain the batches rather than iteratively build a single slice?
 		// If we iteratively build a single slice, we'll spend a lot of time copying elements as the slice grows beyond its capacity.
 		// So instead, we build the slice in one go once we know how many series we have.
 		var streamingSeriesBatches [][]labels.Labels
 		streamingSeriesCount := 0
 
 		for {
+			// XXX: Note that while we free responses' gRPC buffers on error, we don't do the same in case of success,
+			// as the combined response retains references to gRPC buffers.
```

Comment:
Now, thinking twice about this PR: does it make sense to have this extra code optimizing the performance of a non-happy case? 

I think we should fix (or at least evaluate if it's intended) the usage of pooled data instead.

Reply:
The way I'm thinking about it is that this is a first, and easy step, before we fix also the happy path. It's difficult to say how often the `queryIngester` closure will fail in practice.